pkg/strategy/grid2/strategy.go

```diff
@@ -1414,34 +1416,81 @@ func (s *Strategy) recoverGridWithOpenOrdersByScanningTrades(ctx context.Context
 	// 3. get the filled orders from pin-order map
 	filledOrders := pinOrdersFilled.AscendingOrders()
 	numsFilledOrders := len(filledOrders)
-	i := 0
 	if numsFilledOrders == int(expectedOrderNums-openOrdersOnGridNums) {
 		// nums of filled order is the same as Size - 1 - num(open orders)
 	} else if numsFilledOrders == int(expectedOrderNums-openOrdersOnGridNums+1) {
-		i++
+		filledOrders = filledOrders[1:]
 	} else {
 		return fmt.Errorf("not reasonable num of filled orders")
 	}
 
-	// 4. emit the filled orders
+	// 4. verify the grid
+	if err := s.verifyFilledGrid(s.grid.Pins, pinOrdersOpen, filledOrders); err != nil {
+		return errors.Wrapf(err, "verify grid with error")
+	}
+
+	// 5. emit the filled orders
 	activeOrderBook := s.orderExecutor.ActiveMakerOrders()
-	for ; i < numsFilledOrders; i++ {
-		filledOrder := filledOrders[i]
+	for _, filledOrder := range filledOrders {
 		s.logger.Infof("[DEBUG] emit filled order: %s (%s)", filledOrder.String(), filledOrder.UpdateTime)
 		activeOrderBook.EmitFilled(filledOrder)
 	}
 
-	// 5. emit grid ready
+	// 6. emit grid ready
 	s.EmitGridReady()
 
-	// 6. debug and send metrics
+	// 7. debug and send metrics
 	debugGrid(s.logger, grid, s.orderExecutor.ActiveMakerOrders())
```

Comment:
the order book may be not updated yet because the EmitFilled will be delayed, 1 second per trigger.